src/model_theory/terms_and_formulas.lean

```diff
@@ -253,5 +267,91 @@ begin
+∀ (T0 : finset L.sentence), (T0 : L.Theory) ⊆ T → (T0 : L.Theory).is_satisfiable
+
+/-- Given that a theory is satisfiable, selects a model using choice. -/
+def is_satisfiable.some_model {T : L.Theory} (h : T.is_satisfiable) : Type* :=
+  classical.some h
+
+instance is_satisfiable.nonempty_some_model {T : L.Theory} (h : T.is_satisfiable) :
+  nonempty (h.some_model) :=
+classical.some (classical.some_spec h)
+
+noncomputable instance is_satisfiable.inhabited_some_model {T : L.Theory} (h : T.is_satisfiable) :
+  inhabited (h.some_model) := classical.inhabited_of_nonempty'
+
+noncomputable instance is_satisfiable.some_model_structure {T : L.Theory} (h : T.is_satisfiable) :
+  L.Structure (h.some_model) :=
+classical.some (classical.some_spec (classical.some_spec h))
+
+lemma is_satisfiable.some_model_realize_theory {T : L.Theory} (h : T.is_satisfiable) :
+  T.model h.some_model :=
+classical.some_spec (classical.some_spec (classical.some_spec h))
+
+lemma is_satisfiable.of_model {T : L.Theory} (M : Type (max u v)) [n : nonempty M]
+  [S : L.Structure M] (h : T.model M) : T.is_satisfiable :=
+⟨M, n, S, h⟩
+
+lemma is_satisfiable.is_satisfiable_subset {T T' : L.Theory} (h : T'.is_satisfiable) (hs : T ⊆ T') :
+  T.is_satisfiable :=
+⟨h.some_model, h.nonempty_some_model, h.some_model_structure,
+  realize_theory_subset h.some_model_realize_theory hs⟩
+
+lemma is_satisfiable.is_finitely_satisfiable {T : L.Theory} (h : T.is_satisfiable) :
+  T.is_finitely_satisfiable :=
+λ _, h.is_satisfiable_subset
+
+variable {n : ℕ}
+
+/-- Two (bounded) formulas are semantically equivalent over a theory `T` when they have the same
+interpretation in every model of `T`. (This is also known as logical equivalence, which also has a
+proof-theoretic definition.) -/
+def semantically_equivalent (T : L.Theory) (φ ψ : L.bounded_formula α n) : Prop :=
+∀ (M : Type (max u v)) (str : L.Structure M), @model L M str T →
+  @realize_bounded_formula L M str _ _ φ = @realize_bounded_formula L M str _ _ ψ
+
+lemma semantically_equivalent_model {T : L.Theory} {φ ψ : L.bounded_formula α n}
+  {M : Type (max u v)} [str : L.Structure M] (hM : T.model M)
+  (h : T.semantically_equivalent φ ψ) :
+  realize_bounded_formula M φ = realize_bounded_formula M ψ :=
+h M _ hM
+
+lemma semantically_equivalent_some_model {T : L.Theory} {φ ψ : L.bounded_formula α n}
+  (hsat : T.is_satisfiable) (h : T.semantically_equivalent φ ψ) :
+  realize_bounded_formula (hsat.some_model) φ = realize_bounded_formula (hsat.some_model) ψ :=
+h hsat.some_model _ hsat.some_model_realize_theory
+
+/-- Semantic equivalence forms an equivalence relation on formulas. -/
+def semantically_equivalent_setoid (T : L.Theory) : setoid (L.bounded_formula α n) :=
+{ r := semantically_equivalent T,
+  iseqv := ⟨λ φ M str hM, rfl, λ φ ψ se M str hM, (se M str hM).symm,
+    λ φ ψ θ φψ ψθ M str hM, (φψ M str hM).trans (ψθ M str hM)⟩ }
+
+lemma not_not_semantically_equivalent {T : L.Theory} {φ : L.bounded_formula α n} :
```

Comment:
```suggestion
lemma semantically_equivalent_not_not {T : L.Theory} {φ : L.bounded_formula α n} :
```
I think we should encode the conclusion left-to-right (also in lemmas below). If you disagree, feel free to leave them like this.